executor/builder.go

```diff
@@ -1344,53 +1344,50 @@ func (builder *dataReaderBuilder) buildTableReaderFromHandles(goCtx goctx.Contex
 }
 
 func (builder *dataReaderBuilder) buildIndexReaderForIndexJoin(goCtx goctx.Context, v *plan.PhysicalIndexReader,
-	values [][]types.Datum, indexRanges []*ranger.IndexRange, offsetsMap []int) (Executor, error) {
+	values [][]types.Datum, indexRanges []*ranger.IndexRange, keyOff2IdxOff []int) (Executor, error) {
 	e, err := buildNoRangeIndexReader(builder.executorBuilder, v)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
-	// indexRanges is ordered, values is also ordered, and values is used to set value to some equal condition's range.
-	// So the final kvRanges is still ordered.
-	kvRanges := make([]kv.KeyRange, 0, len(indexRanges)*len(values))
-	for _, val := range values {
-		for _, ran := range indexRanges {
-			for i, pos := range offsetsMap {
-				ran.LowVal[pos] = val[i]
-				ran.HighVal[pos] = val[i]
-			}
-		}
-		tmpKvRanges, err1 := indexRangesToKVRanges(e.tableID, e.index.ID, indexRanges)
-		if err1 != nil {
-			return nil, errors.Trace(err)
-		}
-		kvRanges = append(kvRanges, tmpKvRanges...)
+	kvRanges, err := buildKvRangesForIndexJoin(e.tableID, e.index.ID, values, indexRanges, keyOff2IdxOff)
+	if err != nil {
+		return nil, errors.Trace(err)
 	}
 	err = e.open(goCtx, kvRanges)
 	return e, errors.Trace(err)
 }
 
 func (builder *dataReaderBuilder) buildIndexLookUpReaderForIndexJoin(goCtx goctx.Context, v *plan.PhysicalIndexLookUpReader,
-	values [][]types.Datum, indexRanges []*ranger.IndexRange, offsetsMap []int) (Executor, error) {
+	values [][]types.Datum, indexRanges []*ranger.IndexRange, keyOff2IdxOff []int) (Executor, error) {
 	e, err := buildNoRangeIndexLookUpReader(builder.executorBuilder, v)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
-	// indexRanges is ordered, values is also ordered, and values is used to set value to some equal condition's range.
-	// So the final kvRanges is still ordered.
-	kvRanges := make([]kv.KeyRange, 0, len(indexRanges)*len(values))
-	for _, val := range values {
+	kvRanges, err := buildKvRangesForIndexJoin(e.tableID, e.index.ID, values, indexRanges, keyOff2IdxOff)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	err = e.open(goCtx, kvRanges)
+	return e, errors.Trace(err)
+}
+
+// buildKvRangesForIndexJoin builds kv ranges for index join when the inner plan is index scan plan.
+// keyDatums are multiple rows' key datums and is sorted previously. indexRanges are also ordered.
+// So the kvRanges are still ordered.
+func buildKvRangesForIndexJoin(tableID, indexID int64, keyDatums [][]types.Datum, indexRanges []*ranger.IndexRange, keyOff2IdxOff []int) ([]kv.KeyRange, error) {
+	kvRanges := make([]kv.KeyRange, 0, len(indexRanges)*len(keyDatums))
+	for _, val := range keyDatums {
```

Comment:
Make the `keyDatums` in the most inner loop is more efficient when the `indexRanges` and `keyOff2IdxOff` have length 1.

Reply:
This can not make kvRanges ordered.

Reply:
Loop of `KeyDatums` should be the most outside one. Because we iterate the `keyDatums` and then fill the a group of `indexRanges` to produce a group of `kvRanges`.